relays/ethereum/src/ethereum_client.rs

```diff
@@ -418,15 +420,27 @@ impl EthereumHighLevelRpc for EthereumRpcClient {
 /// Substrate headers submitter API.
 #[async_trait]
 trait HeadersSubmitter {
-	/// Returns Ok(true) if not-yet-imported header is incomplete.
-	/// Returns Ok(false) if not-yet-imported header is complete.
+	/// Returns Ok(0) if all given not-yet-imported headers are complete.
+	/// Returns Ok(index != 0) where index is 1-based index of first header that is incomplete.
 	///
-	/// Returns Err(()) if contract has rejected header. This probably means
-	/// that the header is already imported by the contract.
-	async fn is_header_incomplete(&self, header: &QueuedSubstrateHeader) -> Result<bool>;
-
-	/// Submit given header to Ethereum node.
-	async fn submit_header(&mut self, header: QueuedSubstrateHeader) -> Result<()>;
+	/// Returns Err(()) if contract has rejected headers. This means that the contract is
+	/// unable to import first header (e.g. it may already be imported).
+	async fn is_headers_incomplete(
+		&self,
+		header1: &QueuedSubstrateHeader,
+		header2: Option<&QueuedSubstrateHeader>,
+		header3: Option<&QueuedSubstrateHeader>,
+		header4: Option<&QueuedSubstrateHeader>,
```

Comment:
I'm not a fan of this. Couldn't we package this into a collection and process up to four items at a time? We could even make the number of items processed a global constant to make it more flexible if we decide to change the number of headers processed in the future

Reply:
Ok, I've tried && found it worse than passing exactly 4 headers - simply because contract api also accepts headers as 4 separate arguments. So I had to actually pack headers first in `(&QueuedSubstrateHeader, [Option<&QueuedSubstrateHeader>; 3])` then unpack them again to call the contract. With additional indexing it seems worse to me than the current approach.

Reply:
The thing is that it matches the Solidity contract API, IMHO it's better to keep it consistent - pretending on the API level that we can support arbitrary number of headers, while we in fact it's 1-4 will not end up well.
For convenience I'd probably implement a custom type like this:
```
Headers {
 pub header1: QueuedSubstrateHeader,
 pub header2: Option<QueuedSubstrateHeader>,
 pub header3: Option<QueuedSubstrateHeader>,
 pub header4: Option<QueuedSubstrateHeader>,
}
```
and implement some builders or `From` implementations, but it's used in only one place, so I don't feel like it's a strict requirement.

Reply:
> So I had to actually pack headers first in (&QueuedSubstrateHeader, [Option<&QueuedSubstrateHeader>; 3])

Could we not just make everything an `Option`, even though the first header is mandatory? Then we'd be able to have an `[Option<&QueuedSubstrateHeader; MAX_SOLIDITY_HEADERS]` which we can use

Reply:
> For convenience I'd probably implement a custom type like this:
> ```
> Headers {
>  pub header1: QueuedSubstrateHeader,
>  pub header2: Option<QueuedSubstrateHeader>,
>  pub header3: Option<QueuedSubstrateHeader>,
>  pub header4: Option<QueuedSubstrateHeader>,
> }
> ```


This is slightly better, and I'd be okay with this approach as longs as the builders/`From` implementations make it somewhat ergonomic to use